CPDShell/Core/algorithms/knn_algorithm.py

```diff
@@ -0,0 +1,175 @@
+        """
+        Processes a window of data to detect/localize all change points depending on working mode.
+
+        :param window: part of global data for change points analysis.
+        """
+        sample = deque(window)
+        sample_size = len(sample)
+        if sample_size == 0:
+            return
+
+        # Preparing.
+        self.__change_points: list[int] = []
+        self.__change_points_count = 0
+
+        # Building graph.
+        self.__knngraph = knngraph.KNNGraph(window, self.__metric, self.__k)
+        self.__knngraph.build()
+
+        # Examining each point.
+        # Boundaries are always change points.
+        first_point = int(len(window) / 4)
+        last_point = int(len(window) * 0.75)
+
+        for time in range(first_point, last_point):
+            statistics = self.__calculate_statistics_in_point(time, len(window))
+            # print(time, statistics)
+            if self.__check_change_point(statistics):
+                self.__change_points.append(time)
+                self.__change_points_count += 1
+
+    def __calculate_statistics_in_point(self, time: int, window_size: int) -> float:
+        """
+        Calculate the statistics of the KNN graph in specified point.
+
+        :param time: index of point in the given sample to calculate statistics relative to it.
+        :param window_size: size of sample to analyze.
+        """
+        assert self.__knngraph is not None, "Graph should not be None."
+
+        k = self.__k
+        n = window_size
+        n_1 = time
+        n_2 = n - time
+
+        if n <= k:
+            # Unable to analyze sample due to its size.
+            return -k
+
+        h = 4 * (n_1 - 1) * (n_2 - 1) / ((n - 2) * (n - 3))
+
+        sum_1 = (1 / n) * sum(
+            self.__knngraph.check_for_neighbourhood(i, j) * self.__knngraph.check_for_neighbourhood(j, i)
+            for i in range(window_size)
+            for j in range(window_size)
+        )
+
+        sum_2 = (1 / n) * sum(
+            self.__knngraph.check_for_neighbourhood(j, i) * self.__knngraph.check_for_neighbourhood(m, i)
+            for i in range(window_size)
+            for j in range(window_size)
+            for m in range(window_size)
+        )
+
+        expectation = 4 * k * n_1 * (n_2) / (n - 1)
+        variance = (expectation / k) * (h * (sum_1 + k - (2 * k**2 / (n - 1))) + (1 - h) * (sum_2 - k**2))
+        deviation = sqrt(variance)
+
+        permutation: np.array = np.arange(window_size)
+        # np.random.shuffle(permutation) # It seems that random permutation spoils the result
+
+        random_variable_value = self.__calculate_random_variable(permutation, time, window_size)
+
+        if deviation == 0:
+            # if deviation is zero, it likely means that time is 1. This implies that h is 0 and sum_2 = k**2.
+            # In this case we can for sure say that there is no change-point.
+            return -(random_variable_value - expectation)
```

Comment:
Why are we looking specifically at negative difference? Can't we have a random variable deviating from expectation in any direction?